File: solutions/Medium/0547-number-of-provinces/solution.java
// ──────────────────────────────────────────────────
// Problem  : 547. Number of Provinces
// Difficulty: Medium
// Tags     : Depth-First Search, Breadth-First Search, Union-Find, Graph Theory
// Link     : https://leetcode.com/problems/number-of-provinces/
// Runtime  : 6 ms (beats 5%)
// Memory   : 48484000 (beats 23%)
// Language : java
// Copyright: (c) 2026 Shanmukh-dev. All rights reserved.
// Synced by: leetie
// ──────────────────────────────────────────────────

class Solution {
    void dfs(Map<Integer, List<Integer>> adj, int node, int[] visit){
        visit[node] = 1;

        for(int it: adj.get(node)){
            if(visit[it] != 1)
                dfs(adj, it, visit);
        }
    }
    public int findCircleNum(int[][] isConnected) {
        Map<Integer, List<Integer>> adj = new HashMap<>();
        int n = isConnected.length;
        for(int i = 0; i < n; i++){
            adj.put(i, new ArrayList<>());
        }

        for(int i = 0; i < n; i++){
            for(int j = 0; j < n; j++){
                if(isConnected[i][j] == 1 && i != j){
                    adj.get(i).add(j);
                    adj.get(j).add(i);
                }
            }
        }

        int[] visit = new int[n];
        int cnt = 0;
        for(int i = 0; i < n; i++){
            if(visit[i] != 1) {
                dfs(adj, i, visit);
                cnt++;
            }
        }

        return cnt;
    }
}

File: solutions/Medium/0547-number-of-provinces/solution.py
# ──────────────────────────────────────────────────
# Problem  : 547. Number of Provinces
# Difficulty: Medium
# Tags     : Depth-First Search, Breadth-First Search, Union-Find, Graph Theory
# Link     : https://leetcode.com/problems/number-of-provinces/
# Runtime  : N/A (beats 0%)
# Memory   : N/A (beats 0%)
# Language : python3
# Copyright: (c) 2026 Shanmukh-dev. All rights reserved.
# Synced by: leetie
# ──────────────────────────────────────────────────

class Solution:
    def findCircleNum(self, isConnected: List[List[int]]) -> int:
        n = len(isConnected)
        graph = defaultdict(list)

        for i in range(n):
            for j in range(n):
                print(i+1, j+1, isConnected[i][j])
                if i == j:
                    if not graph.get(i+1):
                        graph[i+1] = []

                elif isConnected[i][j] == 1:
                    graph[i+1].append(j+1)

        
        q = deque([1])
        vis = set()

        while q:
            node = q.popleft()
            vis.add(node)
            for i in graph(node):
                



                
        print(graph)